# **Simulation and Modeling Lab**  
**Course Code:** CSE 413  
**Academic Level/Term:** 4/1  
**Credit Hour:** 3  
**Course Teacher:** Audity Ghosh, Lecturer, Department of Computer Science and Engineering, University of Information Technology and Sciences (UITS)  
**Date:** 29 July, 2025  

**Submitted By**  
**Name: Sazzad M Rahman**  
**ID: 0432320005101139**


# Necessary Libraries

In [9]:
import numpy as np
from sympy import Matrix
import matplotlib.pyplot as plt
import seaborn as sns


# Lab Assignment 2 — Unique Matrix Value Investigation
**Student ID:** 1139

For ID **1139**:
- d1 = 9 (last digit)
- d2 = 3 (second-last digit)

According to the assignment,


A =\begin{bmatrix}
d1+2 & d2+1\\
2d1 & d2+2
\end{bmatrix}



In [11]:
# Step 1: Create the personal matrix from student ID 1139
import numpy as np

student_id = "1139"
d1 = int(student_id[-1])
d2 = int(student_id[-2])

A = np.array([
    [d1 + 2, d2 + 1],
    [2 * d1, d2 + 2]
], dtype=float)

print("Student ID:", student_id)
print("d1 =", d1)
print("d2 =", d2)
print("\nMatrix A:")
print(A)


Student ID: 1139
d1 = 9
d2 = 3

Matrix A:
[[11.  4.]
 [18.  5.]]


## Step 2: Analyze Matrix A

Using NumPy, calculate:
- Shape
- Determinant
- Rank
- Eigenvalues
- Inverse Inverse (only if determinant ≠ 0)


In [12]:
# Step 2: Analyze Matrix A
shape_A = A.shape
det_A = np.linalg.det(A)
rank_A = np.linalg.matrix_rank(A)
eigenvalues_A = np.linalg.eigvals(A)

if not np.isclose(det_A, 0):
    inverse_A = np.linalg.inv(A)
else:
    inverse_A = None

print("Shape of A:", shape_A)
print("Determinant of A:", round(det_A, 6))
print("Rank of A:", rank_A)
print("Eigenvalues of A:", np.round(eigenvalues_A, 6))

if inverse_A is not None:
    print("Inverse of A:")
    print(np.round(inverse_A, 6))
else:
    print("A is singular, so it has no inverse.")


Shape of A: (2, 2)
Determinant of A: -17.0
Rank of A: 2
Eigenvalues of A: [17. -1.]
Inverse of A:
[[-0.294118  0.235294]
 [ 1.058824 -0.647059]]


## Step 3: Value Change Experiment

I changed **exactly one element** of Matrix A.

\[
Matrix A:
\begin{bmatrix}
11 & 4\\
 18 & 5
 \end{bmatrix}
\]

The top-left element, `A[0,0]`, was changed from **11 to 12** by adding `+1`.

Therefore,

\[
B =
\begin{bmatrix}
12 & 4\\
18 & 5
\end{bmatrix}
\]


In [13]:
# Step 3: Create Matrix B by changing exactly one value
B = A.copy()

# Change A[0,0] from 11 to 12 by adding +1
B[0, 0] += 1

print("Changed element: A[0,0]")
print("Original value:", A[0, 0])
print("New value:", B[0, 0])
print("\nMatrix B:")
print(B)


Changed element: A[0,0]
Original value: 11.0
New value: 12.0

Matrix B:
[[12.  4.]
 [18.  5.]]


## Step 4: Re-analyze Matrix B


In [14]:
# Step 4: Analyze Matrix B
shape_B = B.shape
det_B = np.linalg.det(B)
rank_B = np.linalg.matrix_rank(B)
eigenvalues_B = np.linalg.eigvals(B)

if not np.isclose(det_B, 0):
    inverse_B = np.linalg.inv(B)
else:
    inverse_B = None

print("Shape of B:", shape_B)
print("Determinant of B:", round(det_B, 6))
print("Rank of B:", rank_B)
print("Eigenvalues of B:", np.round(eigenvalues_B, 6))

if inverse_B is not None:
    print("Inverse of B:")
    print(np.round(inverse_B, 6))
else:
    print("B is singular, so it has no inverse.")


Shape of B: (2, 2)
Determinant of B: -12.0
Rank of B: 2
Eigenvalues of B: [17.67878 -0.67878]
Inverse of B:
[[-0.416667  0.333333]
 [ 1.5      -1.      ]]


# **Comparison & Explanation**
### Answers of the Comparison Questions

**1. How did the determinant change and why?**

The determinant changed from **-17 for A** to **-12 for B**. I increased the top-left element from 11 to 12, and the determinant of a 2×2 matrix depends directly on its entries. For A, the determinant is 11×5 - 4×18 = -17. After changing 11 to 12, it becomes 12×5 - 4×18 = -12. Therefore, the determinant increased by 5.

**2. Did the rank change? Explain.**

No. The rank remained **2** for both matrices. Both determinants are non-zero, which means both 2×2 matrices have two linearly independent rows/columns and therefore have full rank.

**3. How did the eigenvalues respond to the value change?**

The eigenvalues changed because eigenvalues depend on all values of a matrix. Matrix A has eigenvalues **17 and -1**. After increasing one element by 1, Matrix B has eigenvalues approximately **17.678780 and -0.678780**. This shows that even a small change in one matrix element can shift the eigenvalues.

**4. Is B easier or harder to invert than A? Why?**

Both A and B are invertible because both have non-zero determinants. However, B has a determinant with a smaller absolute value: |-12| < |-17|. This means B is  closer to being singular than A. Therefore, B can be considered slightly harder or less numerically stable to invert, although NumPy can invert both matrices without difficulty.
